### PR TITLE
Update wav2vec blog post

### PR DESCRIPTION
**Pull request by patrickvonplaten**

### CODE REVIEW

File: fine-tune-xlsr-wav2vec2.md [hunk 1]
```diff
@@ -24,39 +24,102 @@ thumbnail: /blog/assets/15_fine_tune_wav2vec2/wav2vec2.png
...
+downstream tasks such as speech recognition, speech translation and
+audio classification (*i.e.* diagram on the right below).
+
+![wav2vec2\_structure](https://raw.githubusercontent.com/patrickvonplaten/scientific_images/master/xls_r.png)
+
+XLS-R shows impressive improvements over previous state-of-the-art
+results on both speech recognition, speech translation and
+speaker/language identification, *cf.* with Table 3-6, Table 7-10, and
+Table 11-12 respectively of the official [paper](Fill%20me...).
+TODO(PVP): add link here
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:79]: Remember to update this as well

File: fine-tune-xlsr-wav2vec2.md [hunk 2]
```diff
@@ -24,39 +24,102 @@ thumbnail: /blog/assets/15_fine_tune_wav2vec2/wav2vec2.png
...
+
+Notebook Setup
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:81]: Maybe in the blog post it should not say "notebook". Also it could link to the colab

File: fine-tune-xlsr-wav2vec2.md [hunk 3]
```diff
@@ -24,39 +24,102 @@ thumbnail: /blog/assets/15_fine_tune_wav2vec2/wav2vec2.png
...
 !pip install transformers==4.11.3
+!pip install huggingface_hub==0.1
 !pip install torchaudio
 !pip install librosa
 !pip install jiwer
 ```
 
-Next we strongly suggest to upload your training checkpoints directly to the [🤗 Hub](https://huggingface.co/) while training. The [🤗 Hub](https://huggingface.co/) has integrated version control so you can be sure that no model checkpoint is getting lost during training. 
+We strongly suggest to upload your training checkpoints directly to the
+[🤗 Hub](https://huggingface.co/) while training. The [🤗
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:116]: Nit, for the Hub we try to avoid the emoji. 

```suggestion
[Hugging Face Hub](https://huggingface.co/) while training. The [Hugging Face
```

File: fine-tune-xlsr-wav2vec2.md [hunk 4]
```diff
@@ -66,66 +129,103 @@ notebook_login()
...
+classes to the output transcription.
+
+### Create `Wav2Vec2CTCTokenizer`
+
+A pre-trained XLS-R model maps the speech signal to a sequence of
+context representations as illustrated in the figure above. However, for
+speech recognition the model has to to map this sequence of context
+representations to its corresponding transcription which means that a
+linear layer has to be added on top of the transformer block (shown in
+yellow in the diagram above). This linear layer is used to classifies
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:173]:
```suggestion
yellow in the diagram above). This linear layer is used to classify
```

File: fine-tune-xlsr-wav2vec2.md [hunk 5]
```diff
@@ -66,66 +129,103 @@ notebook_login()
...
+each context representation to a token class analogous how, *e.g.*,
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:174]: This reads a bit difficult and confusing

```suggestion
each context representation to a token class analogous to how
a linear layer is added on top of BERT\'s embeddings
for further classification after pre-training (*cf.* with *\'BERT\'* section of the following [blog
post](https://huggingface.co/blog/warm-starting-encoder-decoder)).
```

File: fine-tune-xlsr-wav2vec2.md [hunk 6]
```diff
@@ -203,106 +315,159 @@ show_random_elements(common_voice_train.remove_columns(["path", "audio"]))
...
+[Merve](https://twitter.com/mervenoyann) was kind enough to take a quick
+look and noted that \"hatted\" characters - like `â` - aren\'t really
+used anymore in Turkish and can be replaced by their \"un-hatted\"
+equivalent, *e.g.* `a`.
+
+This means that we should replace a sentence like
+`"yargı sistemi hâlâ sağlıksız"` to `"yargı sistemi hala sağlıksız"`.
+
+Let\'s write another short mapping function to further simplify the text
+labels. Remember - the simler the text labels, the easier it is for the
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:333]:
```suggestion
labels. Remember, the simpler the text labels, the easier it is for the
```

File: fine-tune-xlsr-wav2vec2.md [hunk 7]
```diff
@@ -203,106 +315,159 @@ show_random_elements(common_voice_train.remove_columns(["path", "audio"]))
...
-
-It is always advantageous to get help from a native speaker of the language you would like to transcribe to verify whether the assumptions you made are sensible, *e.g.* I should have made sure that keeping `'`, but removing other special characters is a sensible choice for Turkish.
-
-To make it clearer that `" "` has its own token class, we give it a more visible character `|`. In addition, we also add an "unknown" token so that the model can later deal with characters not encountered in Common Voice\'s training set.
+Cool, we see that all letters of the alphabet occur in the dataset
+(which is not really surprising) and we also extracted the special
+characters `""` and `'`. Note that we did not exclude those special
+characters because:
+
+The model has to learn to predict when a word is finished or else the
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:432]: The line break above seems off, it was there before due to it being a list, but that was deleted

File: fine-tune-xlsr-wav2vec2.md [hunk 8]
```diff
@@ -312,20 +477,22 @@ with open('vocab.json', 'w') as vocab_file:
...
 
-tokenizer = Wav2Vec2CTCTokenizer("./vocab.json", unk_token="[UNK]", pad_token="[PAD]", word_delimiter_token="|")
+tokenizer = Wav2Vec2CTCTokenizer.from_pretrained("./", unk_token="[UNK]", pad_token="[PAD]", word_delimiter_token="|")
 ```
 
-If one wants to re-use the just created tokenizer with the fine-tuned model of this notebook, it is strongly advised to upload the `tokenizer` to the [🤗 Hub](https://huggingface.co/). Let's call the repo to which we will upload the files
-`"wav2vec2-large-xlsr-turkish-demo-colab"`:
+If one wants to re-use the just created tokenizer with the fine-tuned
+model of this notebook, it is strongly advised to upload the `tokenizer`
+to the [🤗 Hub](https://huggingface.co/). Let\'s call the repo to which
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:491]:
```suggestion
to the [Hugging Face Hub](https://huggingface.co/). Let\'s call the repo to which
```

File: fine-tune-xlsr-wav2vec2.md [hunk 9]
```diff
@@ -334,35 +501,67 @@ and upload the tokenizer to the [🤗 Hub](https://huggingface.co/).
...
-- `sampling_rate`: The sampling rate at which the model is trained on.
-- `padding_value`: For batched inference, shorter inputs need to be padded with a specific value
-- `do_normalize`: Whether the input should be *zero-mean-unit-variance* normalized or not. Usually, speech models perform better when normalizing the input
-- `return_attention_mask`: Whether the model should make use of an `attention_mask` for batched inference. In general, XLSR-Wav2Vec2 models should **always** make use of the `attention_mask`.
+Great, you can see the just created repository under
+`https://huggingface.co/<your-username>/wav2vec2-large-xls-r-300m-tr-colab`
+
+### Create `Wav2Vec2FeatureExtractor`
+
+Speech is a continuous signal and to be treated by computers, it first
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:509]:
```suggestion
Speech is a continuous signal, and, to be treated by computers, it first
```

File: fine-tune-xlsr-wav2vec2.md [hunk 10]
```diff
@@ -334,35 +501,67 @@ and upload the tokenizer to the [🤗 Hub](https://huggingface.co/).
...
+has to be discretized, which is usually called **sampling**. The
+sampling rate hereby plays an important role in that it defines how many
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:511]:
```suggestion
sampling rate hereby plays an important role since it defines how many
```

File: fine-tune-xlsr-wav2vec2.md [hunk 11]
```diff
@@ -334,35 +501,67 @@ and upload the tokenizer to the [🤗 Hub](https://huggingface.co/).
...
+data points of the speech signal are measured per second. Therefore,
+sampling with a higher sampling rate results in a better approximation
+of the *real* speech signal but also necessitates more values per
+second.
+
+A pretrained checkpoint expects its input data to have been sampled more
+or less from the same distribution as the data it was trained on. The
+same speech signals sampled at two different rates have a very different
+distribution, *e.g.*, doubling the sampling rate results in data points
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:520]:
```suggestion
distribution. For example, doubling the sampling rate results in data points
```

File: fine-tune-xlsr-wav2vec2.md [hunk 12]
```diff
@@ -374,76 +573,91 @@ Next, we can prepare the dataset.
...
+            -8.8930130e-05, -3.8027763e-05, -2.9146671e-05], dtype=float32),
+     'path': '/root/.cache/huggingface/datasets/downloads/extracted/05be0c29807a73c9b099873d2f5975dae6d05e9f7d577458a2466ecb9a2b0c6b/cv-corpus-6.1-2020-12-11/tr/clips/common_voice_tr_21921195.mp3',
+     'sampling_rate': 48000}
 ```
 
-Great, we can see that the audio file has automatically been loaded. This is thanks to the new [`"Audio" feature`](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=audio#datasets.Audio) introduced in datasets == 4.13.3, which loads and resamples audio files on-the-fly upon calling.
+Great, we can see that the audio file has automatically been loaded.
+This is thanks to the new [`"Audio"`
+feature](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=audio#datasets.Audio)
+introduced in `datasets == 4.13.3`, which loads and resamples audio
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:605]: Maybe this should link to the release notes

File: fine-tune-xlsr-wav2vec2.md [hunk 13]
```diff
@@ -454,77 +668,127 @@ print("Sampling rate:", common_voice_train[rand_int]["audio"]["sampling_rate"])
...
- seconds out of the training dataset.
+**Note**: Currently `datasets` make use of
+[`torchaudio`](https://pytorch.org/audio/stable/index.html) and
+[`librosa`](https://librosa.org/doc/latest/index.html) for audio loading
+and resampling. If you wish to implement your own costumized data
+loading/sampling, feel free to just make use of the `"path"` column
+instead and disregard the `"audio"` column.
+
+Long input sequences require a lot of memory. XLS-R is based on
+`self-attention` the memory requirement scales quadratically with the
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:733]:
```suggestion
`self-attention`. The memory requirement scales quadratically with the
```

File: fine-tune-xlsr-wav2vec2.md [hunk 14]
```diff
@@ -640,52 +884,79 @@ def compute_metrics(pred):
...
+[docs](https://huggingface.co/transformers/master/main_classes/trainer.html?highlight=trainer#trainingarguments).
 
-During training, a checkpoint will be uploaded asynchronously to the Hub every 400 training steps. It allows you to also play around with the demo widget even while your model is still training.
+During training, a checkpoint will be uploaded asynchronously to the hub
+every 400 training steps. It allows you to also play around with the
+demo widget even while your model is still training.
 
-**Note**: If one does not want to upload the model checkpoints to the Hub, simply set `push_to_hub=False`.
+**Note**: If one does not want to upload the model checkpoints to the
+hub, simply set `push_to_hub=False`.
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:959]:
```suggestion
Hub, simply set `push_to_hub=False`.
```

File: fine-tune-xlsr-wav2vec2.md [hunk 15]
```diff
@@ -640,52 +884,79 @@ def compute_metrics(pred):
...
+    fine-tuning has become stable. Note that those parameters strongly
+    depend on the Common Voice dataset and might be suboptimal for other
+    speech datasets.
 
-For more explanations on other parameters, one can take a look at the [docs](https://huggingface.co/transformers/master/main_classes/trainer.html?highlight=trainer#trainingarguments).
+For more explanations on other parameters, one can take a look at the
+[docs](https://huggingface.co/transformers/master/main_classes/trainer.html?highlight=trainer#trainingarguments).
 
-During training, a checkpoint will be uploaded asynchronously to the Hub every 400 training steps. It allows you to also play around with the demo widget even while your model is still training.
+During training, a checkpoint will be uploaded asynchronously to the hub
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:954]:
```suggestion
During training, a checkpoint will be uploaded asynchronously to the Hub
```

File: fine-tune-xlsr-wav2vec2.md [hunk 16]
```diff
@@ -728,89 +999,104 @@ trainer = Trainer(
...
+| 0.0997        | 18.35 | 2000 | 0.4567          | 0.3696 |
+| 0.0784        | 22.02 | 2400 | 0.4193          | 0.3442 |
+| 0.0633        | 25.69 | 2800 | 0.4153          | 0.3347 |
+| 0.0498        | 29.36 | 3200 | 0.4077          | 0.3195 |
 
-The training loss goes down and we can see that the WER on the test set also improves nicely. Because this notebook is just for demonstration purposes, we can stop here.
+The training loss and validation WER go down nicely.
 
-You can now upload the result of the training to the Hub, just execute this instruction:
+You can now upload the result of the training to the 🤗 Hub, just execute
```
Comment (osanseviero) [inline fine-tune-xlsr-wav2vec2.md:1048]:
```suggestion
You can now upload the result of the training to the Hub, just execute
```

**Review by osanseviero — APPROVED** [review]
Super nice article!

I left a couple of minor suggestions, nits and typos to improve readability, but it is an excellent read!

File: fine-tune-xlsr-wav2vec2.md [hunk 17]
```diff
@@ -4,11 +4,11 @@ thumbnail: /blog/assets/15_fine_tune_wav2vec2/wav2vec2.png
 ---
 
 <h1>
-    Fine-Tune XLSR-Wav2Vec2 for low-resource ASR with 🤗 Transformers
+    Fine-tuning XLS-R for Multi-Lingual ASR with 🤗 Transformers
 </h1>
 
 <div class="blog-metadata">
-    <small>Published March 12, 2021.</small>
+    <small>Published November 15, 2021.</small>
```
Comment (julien-c) [inline fine-tune-xlsr-wav2vec2.md:11]:
```suggestion
    <small>Updated November 15, 2021. Originally published March 12, 2021.</small>
```

**Comment by osanseviero** [comment]
I saw  assets/xlsr_wav2vec2.png  is quite large (+2MB). Could you compress it please?

**Comment by patrickvonplaten** [comment]
> I saw assets/xlsr_wav2vec2.png is quite large (+2MB). Could you compress it please?

It doesn't render well when I compress it too much. Also we have already over 100MB of images in the repo - think we should start by compressing the other ones first or even move the whole image folder out of the repo.

**Comment by osanseviero** [comment]
> It doesn't render well when I compress it too much. Also we have already over 100MB of images in the repo - think we should start by compressing the other ones first or even move the whole image folder out of the repo.

The problem is not with the repo size. The problem is that when users open the blog post, it will be loading huge images. This will translate to slow experience (and costly for mobile users). If you really want to keep this it's ok I guess since it's 1 image, just wanted to explain the reasoning since we've been avoiding it in the last few months.

**Comment by julien-c** [comment]
due to how git works, compressing the old ones won't make the repo any smaller anyways @patrickvonplaten =)

**Comment by patrickvonplaten** [comment]
Down to 1.4MB, but below that the quality does suffer so I'll leave it
